Add CLI task to get/set settings, refs #13604

mcantelon opened the pull request:
Added CLI task to get or set settings, optionally to or from files.

mcantelon committed:
Add CLI task to get/set settings, refs #13604

Added CLI task to get or set settings, optionally to or from files.

djjuhasz commented on lib/task/tools/settingsTask.class.php:
```diff
@@ -0,0 +1,351 @@
+
+        $this->validateOptions($arguments, $options);
+
+        $this->displayCurrentSettings();
```
When I run the task I always get the full list of settings, e.g.

```
[dev/issue-13604-settings-task][~/atom]$ symfony tools:setting get version
------------------------------------------------------------------------------------
Name                                                              Scope
------------------------------------------------------------------------------------
_csrf_token                                                       element_visibility
accession_counter
accession_mask
accession_mask_enabled
actor                                                             default_template
...
```

Is this intentional? It makes it hard to actually see the value for the selected setting.

djjuhasz replied:
The "_csrf_token" setting is due to a bug (https://projects.artefactual.com/issues/13613) I just discovered while testing this task. 🤦

jraddaoui replied:
Oh, good catch @djjuhasz!

mcantelon replied:
Weird one, that "_csrf_token". 😨 

Yeah, I added that display of a list of settings to make the settings easier to work with, but displaying it the way I'm displaying it - with every command - is kludge-y. I'll take it out for now... I could make a new CLI task to display settings if need be. That'd be less weird.

jraddaoui commented:
Nice one @mcantelon!

@mamedin recently asked for something similar to get setting and config values, I created this quick "tools:run" task ...

```php
<?php

$config = sfConfig::getAll();
$config['factories'] = sfFactoryConfigHandler::getConfiguration(
    $this->configuration->getConfigPaths('config/factories.yml')
);

echo json_encode($config, JSON_PRETTY_PRINT) . "\n";
```

I think it could be a good addition to include a way to get the values from the factories config, maybe not in this task or without a set option. And also to add an option to output as JSON when printing multiple values, so you can then pipe the output a bit easily ...

`php symfony tools:run output_config_json.php | jq -r '.factories.user.param.timeout'`

mcantelon commented:
Thanks @radda! Good thinking... I've kind of shoehorned settings display into this task so I think I'm going to remove it and add the logic to a new task. Adding display of factory settings makes sense, too, so I'll add a Redmine issue to describe this new task (and specify that it should have JSON output too).

djjuhasz commented on lib/task/tools/settingsTask.class.php:
```diff
@@ -0,0 +1,351 @@
+                'Scope of setting to get or set',
+                null
+            ),
+            new sfCommandOption(
```
Opinion: I think it's more intuitive to make the 'value' an optional argument, so the command can be called like `symfony tools:setting set [name] [value]`

djjuhasz commented on lib/task/tools/settingsTask.class.php:
```diff
@@ -0,0 +1,351 @@
+        ]);
+
+        $this->namespace = 'tools';
+        $this->name = 'setting';
```
Nitpick: I think it's better for the class name to be `SettingTask` or to make `$this->name = 'settings';` for consistency.

mcantelon replied:
Good call. Changing the name to "settings" makes sense.

djjuhasz reviewed (approved):
@mcantelon a few comments but no blockers.  I'm impressed that you simulated the database interactions for the tests. :)

mcantelon commented:
@djjuhasz I've updated the PR and did a bit of minor refactoring, fixed some errors I found, added a couple of new tests, and got rid of the `to-file` and `from-file` options (replacing them with a single `file` option).

(Oh you approved it earlier... thanks! I'll merge it!)
